### DISCUSSION TITLE
Request Forwarding - How to forward to a Lua Language server?

### DISCUSSION THREAD

**Discussion by russell-lew** [microsoft/vscode-discussions#72] [Extension Development QnA] [answered]
Hi everyone, I've also posted this question [here](https://github.com/microsoft/vscode-extension-samples/issues/681), I'm also sharing here to look for some answers.

I am working on a language server extension with embedded Lua code which I want to get Intellisense support on. My extension does not have any language features for Lua. I have decided to use request forwarding to achieve Intellisense.

I played around with the [lsp-embedded-request-forwarding](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-request-forwarding) sample and tried using using completions for different languages by changing the following code:

line 36: updating the slicing to remove .js extensions instead of .css
```
	workspace.registerTextDocumentContentProvider('embedded-content', {
		provideTextDocumentContent: uri => {
			const originalUri = uri.path.slice(1).slice(0, -3);
			const decodedUri = decodeURIComponent(originalUri);
			return virtualDocumentContents.get(decodedUri);
		}
	});
```

line 56: changing the authority and language extension
```
				const vdocUriString = `embedded-content://js/${encodeURIComponent(
					originalUri
				)}.js`;
```
Doing so allowed me to get Javascript completions within the <style> tags (just for testing's sake). However, when I try it with Python and Lua, I was unable to get any completion results when typing in their respective languages.

The [VSCode documentation](https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding) explains that "request forwarding sends the request back to VS Code to query all language servers and forward their responses".

I would like to clarify the scope of 'all language servers' - does that mean all third party ones available on the marketplace as well?

Initially, I thought it could be limited to language servers developed by Microsoft. But since I could not get it to work on Python, it is not behaving the way I expected.

Which language servers and languages are supported by this request forwarding feature? Is it possible to do request forwarding to marketplace language servers such as Lua and how do I do it?

**Comment by gjsjohnmurray** [comment, marked as answer]
Maybe you are being bitten by https://github.com/microsoft/vscode-extension-samples/issues/682 for which I have submitted a PR.

**Reply by russell-lew** [reply]
Disabling the URI encoding just like in the PR did not seem to fix my issue. 

I have identified the problem to be extension specific - request forwarding can receive a response when I installed a different Lua language server extension. 

More specifically, the (extension by Sumneko) [https://github.com/sumneko/lua-language-server] does not send a response through request forwarding, but when I installed the (extension by Keyring)[https://github.com/keyring/vsc-lua], completion suggestions are returned as expected.

**Reply by gjsjohnmurray** [reply]
Perhaps the Sumneko extension makes assumptions about document URIs and can't handle the ones that the request forwarding technique relies on.

**Comment by Warda8841** [comment]
Hi
